### PR TITLE
Move cache related code out of compiler class

### PR DESCRIPTION
**Pull request by s-perron**
We want to move the code that deals with the cache out of the compiler class and into the CacheAccessor class.

I have this divided into multiple commits.  You can look at each one on its own.  I can sqash them once everything is approved.  I can also make each one a separate PR it you other would like.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 13ccf57279b3f5e25efcb957a50faa5adbb98e4f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 9786fa8c34df6d9baddeff40d331106799fcbb07</li>
<li>SPVGEN: dev 5bf4618ac00d4675f57ca5b571a485bbc4d4fa33</li>
<li>XGL: dev 9b77f6008d4316922cde45baf39d079b4d328893</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1b94c3bf02ef6b638e199aa5c41d2fa31ea5455d</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225557)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 30786/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25606/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30785/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25607/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 30917/56381 (54.8%)</li>
<li>Failed: 0/56381 (0.0%)</li>
<li>Not Supported: 25464/56381 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34387/56392 (61.0%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 22005/56392 (39.0%)</li>
</ul></details>


</ul></details>

File: llpc/util/llpcCacheAccessor.cpp [hunk 1]
```diff
@@ -60,24 +65,108 @@ CacheAccessor::CacheAccessor(Context *context, MetroHash::Hash &cacheHash, Compi
...
+}
+
+// =====================================================================================================================
+// Looks for the given hash in the shader caches and sets the cache accessor state with the results.
+//
+// @param hash : The hash to look up.
+void CacheAccessor::lookUpInShaderCaches(MetroHash::Hash &hash) {
+  ShaderCache *applicationCache = static_cast<ShaderCache *>(getApplicationShaderCache());
+  ShaderCache *internalCache = static_cast<ShaderCache *>(getInternalShaderCache());
+  bool usingApplicationCache = applicationCache && llvm::cl::ShaderCacheMode != ShaderCacheForceInternalCacheOnDisk;
```
Comment (amdrexu) [inline llpc/util/llpcCacheAccessor.cpp:127]: Remove "llvm::" prefix

File: llpc/context/llpcCompiler.cpp [hunk 2]
```diff
@@ -1500,7 +1500,7 @@ Result Compiler::BuildGraphicsPipeline(const GraphicsPipelineBuildInfo *pipeline
 
   llvm::Optional<CacheAccessor> cacheAccessor;
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.cpp:1501]: Remove "llvm::" prefix

File: llpc/util/llpcCacheAccessor.cpp [hunk 3]
```diff
@@ -60,24 +65,108 @@ CacheAccessor::CacheAccessor(Context *context, MetroHash::Hash &cacheHash, Compi
...
+  lookUpInICaches(hash);
+  if (m_cacheResult != Result::Success)
+    lookUpInShaderCaches(hash);
+}
+
+// =====================================================================================================================
+// Looks for the given hash in the ICaches and sets the cache accessor state with the results.
+//
+// @param hash : The hash to look up.
+void CacheAccessor::lookUpInICaches(const MetroHash::Hash &hash) {
```
Comment (amdrexu) [inline llpc/util/llpcCacheAccessor.cpp:85]: lookUpInICaches -> lookUpInCaches
I think using ICaches in the function name is somewhat strange.

File: llpc/util/llpcCacheAccessor.cpp [hunk 4]
```diff
@@ -60,24 +65,108 @@ CacheAccessor::CacheAccessor(Context *context, MetroHash::Hash &cacheHash, Compi
...
+// will be allocated if there is a cache miss and allocateOnMiss is true.
+//
+// @param hash : The hash to look up.
+// @param allocateOnMiss : Will add an entry to the cache on a miss if true.
+// @param cache : The cache in whith to look.
+bool CacheAccessor::lookUpInShaderCache(const MetroHash::Hash &hash, bool allocateOnMiss, ShaderCache *cache) {
+  CacheEntryHandle currentEntry;
+  ShaderEntryState cacheEntryState = cache->findShader(hash, allocateOnMiss, &currentEntry);
+  if (cacheEntryState == ShaderEntryState::Ready) {
+    Result r = cache->retrieveShader(currentEntry, &m_elf.pCode, &m_elf.codeSize);
```
Comment (amdrexu) [inline llpc/util/llpcCacheAccessor.cpp:158]: r -> result

File: llpc/util/llpcCacheAccessor.cpp [hunk 5]
```diff
@@ -102,4 +191,23 @@ void CacheAccessor::setElfInCache(BinaryData elf) {
   }
 }
 
+// =====================================================================================================================
+// Updates the entry in the shader cache, if there is one, for this access.
+//
+// @param elf : The binary encoding of the elf to place in the cache.
+void CacheAccessor::updateShaderCache(BinaryData &elf) {
+  ShaderCache *shader_cache = m_shaderCache;
```
Comment (amdrexu) [inline llpc/util/llpcCacheAccessor.cpp:199]: shader_cache -> shaderCahe

**Comment by s-perron** [comment]
@amdrexu I believe I have addressed all of the issues.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7278dedc6129c85d839e77d88c58ec15104c9458
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 9786fa8c34df6d9baddeff40d331106799fcbb07</li>
<li>SPVGEN: dev 5bf4618ac00d4675f57ca5b571a485bbc4d4fa33</li>
<li>XGL: dev 9b77f6008d4316922cde45baf39d079b4d328893</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1b94c3bf02ef6b638e199aa5c41d2fa31ea5455d</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225557)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 30786/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25606/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30785/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25607/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 30917/56381 (54.8%)</li>
<li>Failed: 0/56381 (0.0%)</li>
<li>Not Supported: 25464/56381 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34387/56392 (61.0%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 22005/56392 (39.0%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Comment by s-perron** [comment]
Rebased on top of latest dev.  No changes from previous review.

**Review by amdrexu — APPROVED** [review]
I think it is good to go.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0f2f9c437051e1d3844a1d25e94cd46c63559eec
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 9786fa8c34df6d9baddeff40d331106799fcbb07</li>
<li>SPVGEN: dev 5bf4618ac00d4675f57ca5b571a485bbc4d4fa33</li>
<li>XGL: dev 9b77f6008d4316922cde45baf39d079b4d328893</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1b94c3bf02ef6b638e199aa5c41d2fa31ea5455d</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/225557)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 30786/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25606/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30784/56392 (54.6%)</li>
<li><details><summary>Failed: 1/56392 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.pipeline_barrier.all.1048576_vertex_buffer_stride_2
Stack trace: Script:
13:HostMemoryAccess Result differs from reference, Expected: 0x0000008c, Got: 0x00000050, At offset: 1007588

```
</p></details></li>

<li>Not Supported: 25607/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 30917/56381 (54.8%)</li>
<li>Failed: 0/56381 (0.0%)</li>
<li>Not Supported: 25464/56381 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34387/56392 (61.0%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 22005/56392 (39.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**